File: 61.rotate-list.cpp
/*
 * @lc app=leetcode id=61 lang=cpp
 *
 * [61] Rotate List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

#define DEBUG 0
#if DEBUG
#include <stdio.h>
#include <algorithm>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};
#endif

// Input: head = [1,2,3,4,5], k = 2
// Output: [4,5,1,2,3]
// 1-> 2-> 3-> 4-> 5
// 5-> 1-> 2-> 3-> 4
// 4-> 5-> 1-> 2-> 3

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        int move = k;
        int count = 0;

        if (head == NULL) {
            return head;
        }

        ListNode* tmp = head;
        while(tmp) {
            tmp = tmp->next;
            count++;
        }

        move = k%count;
        while (move) {
            // printf("move=%d\n", move);
            ListNode* cur = head;
            ListNode* pre = NULL;
            ListNode* nxt = NULL;
            while (cur) {
                // printf("cur:%d, count:%d\n", cur->val, count);
                nxt = cur->next;

                if (nxt == NULL) {
                    cur->next = head;
                    pre->next = NULL;
                    head = cur;
                    break;
                }
                pre = cur;
                cur = cur->next;
            }
            move--;
        }
        return head;
    }
};
// @lc code=end

#if DEBUG
int main(int argc, char* argv[])
{
    ListNode *l1_node5 = new ListNode(5);
    ListNode *l1_node4 = new ListNode(4, l1_node5);
    ListNode *l1_node3 = new ListNode(3, l1_node4);
    ListNode *l1_node2 = new ListNode(2, l1_node3);
    ListNode *l1_node1 = new ListNode(1, l1_node2);

    int k = 2;

    for (ListNode *tmp = l1_node1; tmp != NULL; tmp=tmp->next)
    {
        printf("l1 %d\n", tmp->val);
    }

    Solution solution;
    ListNode *result = solution.rotateRight(l1_node1, k);

    for (ListNode *tmp = result; tmp != NULL; tmp=tmp->next)
    {
        printf("result %d\n", tmp->val);
    }



    printf("end\n");
    return 0;
}
#endif

File: 61.rotate-list.py
#
# @lc app=leetcode id=61 lang=python3
#
# [61] Rotate List
#
# https://leetcode.com/problems/rotate-list/description/
#
# algorithms
# Medium (34.89%)
# Likes:    6212
# Dislikes: 1306
# Total Accepted:    624K
# Total Submissions: 1.8M
# Testcase Example:  '[1,2,3,4,5]\n2'
#
# Given the head of a linked list, rotate the list to the right by k places.
#
#
# Example 1:
#
#
# Input: head = [1,2,3,4,5], k = 2
# Output: [4,5,1,2,3]
#
#
# Example 2:
#
#
# Input: head = [0,1,2], k = 4
# Output: [2,0,1]
#
#
#
# Constraints:
#
#
# The number of nodes in the list is in the range [0, 500].
# -100 <= Node.val <= 100
# 0 <= k <= 2 * 10^9
#
#
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if head == None or head.next == None:
            return head

        temp = head
        nodeCnt = 1
        while temp.next != None:
            temp = temp.next
            nodeCnt += 1

        lastNode = temp
        # print(nodeCnt)
        # print(k)
        lastIdxAfterRotated = (nodeCnt - k) % nodeCnt
        # print(lastIdxAfterRotated)

        lastNode.next = head

        while lastIdxAfterRotated:
            lastNode = lastNode.next
            lastIdxAfterRotated -= 1

        head = lastNode.next
        lastNode.next = None

        return head

# @lc code=end

